### PR TITLE
experimental: Allow json in template literals

### LINKED ISSUE #3872: Stringify JSON variable when used inside template string
As a user, want to be able to pass JSON data returned from any API endpoint to client-side javascript. When I can use template string in expression for this I get several options:

- [x] Using JSON inside html embed script as js variable

  ```js
  `<script type="module">const myData = ${myData};</script>`
  ```

- [x] Using JSON inside html embed script as text content

  ```js
  `<script type="application/json">${myData}</script>`
  ```

-  ❌  Using JSON inside html embed as attribute _(@istarkov Can't be valid - example `<div data-json="{"hello":"world"}">attr-here</div>` here should be special case i.e. `{&quot;hello&quot;:&quot;world&quot;}` but it's not acceptable in all other cases)_

  ```js
  `<div data-json='${myData}'></script>`
  ```

- [x]  Using Text component and binding text content

### PR DESCRIPTION
**Pull request by istarkov**
## Description
closes #3872

- [x] - Builder support
- [x] - Publish support

## Steps for reproduction

Example
https://add-json.staging.webstudio.is/builder/8160e16d-5ab4-4249-86de-0c98a64526a3

https://test-json-ki9yj.wstd.work/

## Code Review

- [ ] hi @kof, I need you to do
  - conceptual review (architecture, feature-correctness)
  - detailed review (read every line)
  - test it on preview

## Before requesting a review

- [ ] made a self-review
- [ ] added inline comments where things may be not obvious (the "why", not "what")

## Before merging

- [ ] tested locally and on preview environment (preview dev login: 5de6)
- [ ] updated [test cases](https://github.com/webstudio-is/webstudio/blob/main/apps/builder/docs/test-cases.md) document
- [ ] added tests
- [ ] if any new env variables are added, added them to `.env` file

### CODE REVIEW

File: apps/builder/app/shared/nano-states/props.ts [hunk 1]
```diff
@@ -200,6 +200,34 @@ const $loaderVariableValues = computed(
   }
 );
 
+const createJsonStringifyProxy = <T extends object>(target: T): T => {
```
Comment (istarkov) [inline apps/builder/app/shared/nano-states/props.ts:203]: 99% this will go to sdk

**Comment by kof** [comment]
- [ ] please don't forget to add docs in the relevant places, @johnsicili can help with that ... its a very important feature that needs to be well visible

File: apps/builder/app/builder/shared/binding-popover.test.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import { expect, test } from "@jest/globals";
+import { evaluateExpressionWithinScope } from "./binding-popover";
+import { encodeDataSourceVariable } from "@webstudio-is/sdk";
+
+test("evaluateExpressionWithinScope works", () => {
+  const variableName = "jsonVariable";
+  const encVariableName = encodeDataSourceVariable(variableName);
+  const variableValue = 1;
+  const variables = { [encVariableName]: variableValue };
+  const expression = `${encVariableName} + ${encVariableName}`;
```
Comment (TrySound) [inline apps/builder/app/builder/shared/binding-popover.test.ts:10]: I prefer to inline everything to make test readable

File: packages/sdk/src/to-string.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+      if (typeof value === "object" && value !== null) {
+        return createJsonStringifyProxy(value);
+      }
+
+      return value;
+    },
+  });
+};
+
+export const isPlainObject = (obj: unknown): obj is object => {
```
Comment (kof) [inline packages/sdk/src/to-string.ts:21]: as an idea call obj - value, so you don't need to abbreviate

File: apps/builder/app/builder/shared/binding-popover.test.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+test("evaluateExpressionWithinScope works", () => {
+  const variableName = "jsonVariable";
+  const encVariableName = encodeDataSourceVariable(variableName);
+  const variableValue = 1;
+
+  expect(
+    evaluateExpressionWithinScope(`${encVariableName} + ${encVariableName}`, {
+      [encVariableName]: variableValue,
+    })
```
Comment (TrySound) [inline apps/builder/app/builder/shared/binding-popover.test.ts:13]: imo would be easier to understand the test with variables inlined
```
   evaluateExpressionWithinScope(`$ws$dataSource$jsonVariable + $ws$dataSource$jsonVariable`, {
       '$ws$dataSource$jsonVariable': 1,
     })
```

File: packages/react-sdk/src/context.tsx [hunk 5]
```diff
@@ -49,5 +50,13 @@ export const ReactSdkContext = createContext<
...
 export const useResource = (name: string) => {
   const { resources } = useContext(ReactSdkContext);
-  return resources[name];
+  const resource = useMemo(
+    () =>
+      isPlainObject(resources[name])
+        ? createJsonStringifyProxy(resources[name])
+        : resources[name],
+    [resources, name]
+  );
```
Comment (TrySound) [inline packages/react-sdk/src/context.tsx:59]: Maybe move resources[name] outside of useMemo?

**Review by TrySound — APPROVED** [review]

File: packages/react-sdk/src/context.tsx (hunk 5 above)
Comment (istarkov) [reply packages/react-sdk/src/context.tsx:59]: Done

**Pull request merged**